File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.py
def solution(num):
    stack = []
    l = len(num)
    ans = [-1]*l
    for i in range(l-1,-1,-1):
        while stack and stack[-1] <= num[i]:
            stack.pop()
        if not stack: pass
        else: ans[i] = stack[-1]
        stack.append(num[i])
    return ans

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.kt
class Solution {
    fun solution(numbers: IntArray): IntArray {
        val ans = Array(numbers.size){ -1 }
        val stack = mutableListOf<Int>()
        for (i in numbers.indices.reversed()) {
            while (stack.size > 0 && stack[stack.lastIndex] <= numbers[i]) {
                stack.removeLast()
            }
            if (stack.size > 0) ans[i] = stack[stack.lastIndex]
            stack.add(numbers[i])
        }
        return ans.toIntArray()
    }
}